### ISSUE TITLE
Kinematic constraint between a gripper and a target object

### ISSUE THREAD

**Issue by harryzhangOG** [haosulab/SAPIEN#49]
**Is your feature request related to a problem? Please describe.**
Hi, is there a way to implement a "sticky gripper" in maniskill? i.e. create a kinematic constraint when making contacts? Thanks.

**Describe the solution you'd like**
I think all I need to do is to create a fixed link between the target joint and the gripper finger?
In ManiSkill, here is what we did:
```Python
parent = env.agent.finger1_joint.articulation.get_builder().create_link_builder()
child = env.target_joint.articulation.get_builder().create_link_builder(parent)
child.set_joint_properties('fixed', limits=[], pose_in_parent=sapien.core.Pose(p=[0, 0, 0], q = [0, 0, 0, 0]), pose_in_child=sapien.core.Pose(p=[0, 0, 0], q = [0, 0, 0, 0]),\
                               friction = 1, damping = 1)
```
but it had no visible effect on the environment

**Additional context**
basically it is a gripper that upon contacting the object, it creates a kinematic constraint that never lets go of the object (analogous to a suction gripper). The core question is how to create a fixed kinematic constraint between the gripper and the target.

**Comment by fbxiang** [comment]
I would say what you seem to try to achieve has nothing to with articulations. What you need is a drive.
First you call `scene.create_drive` [here](https://github.com/haosulab/SAPIEN/blob/master/python/py_package/core/pysapien/__init__.pyi#L1423). The actor arguments are the actors you want to stick together. The pose arguments are the pose of the "sticky" joint in the frame of actor1 and actor 2 respectively.

Next, you just want to use large drive stiffness and reasonable damping to pull them together. The drive signature can be found [here](https://github.com/haosulab/SAPIEN/blob/master/python/py_package/core/pysapien/__init__.pyi#L628-L645)

**Comment by harryzhangOG** [comment]
Thanks. May I ask how to access the actor object from an articulation object (env.agent.robot)? Thanks.

**Comment by fbxiang** [comment]
`articulation.get_links`

**Comment by harryzhangOG** [comment]
Thanks. It is working now.

**Comment by harryzhangOG** [comment]
Hi, I implemented the constraint based on your suggestion and it worked for the drawer environment, but for the door environment, it is having some weird visualization issues. Please see the video below:

https://user-images.githubusercontent.com/43732483/139596419-996e9d71-8e4c-4909-8f30-0e1b376423e3.mov

The code I wrote was:
```Python
if phase_counter == 50:
            scene = env._scene
            dr = scene.create_drive(
                env.agent.robot.get_links()[1],
                env.agent.robot.get_links()[1].get_cmass_local_pose(),
                env.target_link,
                env.target_link.get_cmass_local_pose(),
            )
            dr.set_x_properties(stiffness=20, damping=0.5)
            dr.set_y_properties(stiffness=20, damping=0.5)
            dr.set_z_properties(stiffness=20, damping=0.5)
```
Please advise on where we went wrong. Thanks!

**Comment by fbxiang** [comment]
It looks like you are only setting translation constraint but not rotation ones. I suggest adding them and setting the stiffness to a larger value.

**Comment by harryzhangOG** [comment]
Thanks. That did improve the visualization. May I confirm what would be the poses of the two actors to be passed into the create_drive method? Should I just pass them in once at the beginning or iteratively track the poses and call create_drive every time step? Thanks.

**Comment by fbxiang** [comment]
A drive is just a removable joint. The poses are the pose of the joint in the frame of the 2 actors. You can decide your own implementation, but I do expect creating and destroying drives to take a bit of time.

**Comment by harryzhangOG** [comment]
Hi. I did iterative creation and deletion of the drive and I think the translation constraint is working fine now. However, may I ask how to do the angular constraints too? I see there are three set_properties method for angular movements: set_x_twist, set_yz_swing and set_slerp. Which one should I set? Thanks.

**Comment by fbxiang** [comment]
Twist-swing drives the Euler angles and slerp drives the quaternion. You can only use 1 system so you can choose based on your needs.
